### PR TITLE
feat(python): expose efficient iterator over `DataFrame` slices

### PR DESCRIPTION
**Pull request by alexander-beedie**
(Take two 🎬)

The previous effort was clearly (oops :) at the wrong level, but I couldn't get this pattern out of my head; while the functionality and implementation is obviously straightforward, it has value as a first-class method.

Yielding zero-copy slices as frames that can be further operated on by the caller allows for simple/efficient iteration over batches of _anything_ that `DataFrame` supports.

**Example**:

Redoing the previous example, RecordBatch iteration would now represent a trivial usage of `iterslices()` -

```python
from datetime import date
import polars as pl

df = pl.DataFrame(
    {
        "x": range(17_500),
        "y": date(2023,1,1),
        "z": "abcdefghijklm",
    }
)

for frame in df.iterslices(n_rows=15_000):
    record_batch = frame.to_arrow().to_batches()[0]
```

### CODE REVIEW

**Comment by erinov1** [comment]
Not directly related to this PR, but another useful functionality would be to slice a `LazyFrame`/`DataFrame` using a second frame of offsets/length. For example:

```python
df = pl.DataFrame({"A": [1, 2, 3, 4, 5]}).lazy()
slices = pl.DataFrame({"offset": [0, 3], "length": [2, 2]})
(
    df
    .with_context(slices.lazy())
    .slice(pl.col("offset"), pl.col("length"))
    .select(pl.col("A").max())
    .collect()
)
``` 
would return the max of `A` over the slice (0, 2) and (3, 2), namely
```python
pl.DataFrame({"A": [2, 5]})
```
Currently I'm doing this via
```python
sub_dfs = []
for n in slices.height:
    sub_dfs.append(
        (
            df
            .with_context(slices.slice(n,1))
            .slice(pl.col("offset"), pl.col("length"))
            .select(pl.col("A").max())
        )
    )

pl.collect_all(sub_dfs)
```
but this does not feel very ergonomic.

File: py-polars/polars/internals/dataframe/frame.py [hunk 1]
```diff
@@ -6805,6 +6805,57 @@ def iterrows(
             for i in range(self.height):
                 yield self.row(i)
 
+    def iterslices(self, n_rows: int = 10_000) -> Iterator[DataFrame]:
```
Comment (ritchie46) [inline py-polars/polars/internals/dataframe/frame.py:6808]: Can we name this `iter_slices`?

Comment (alexander-beedie) [reply py-polars/polars/internals/dataframe/frame.py:6808]: Sure, though that doesn't really match-up with `iterrows`; should we change both for `0.16`? I agree it looks cleaner to have a little clarifying visual space between the `iter` and _what_ is being iterated over ;)

* `iter_rows`
* `iter_slices`

Comment (ritchie46) [reply py-polars/polars/internals/dataframe/frame.py:6808]: Yes, let's do that with `iter_rows` as well.

I hope to release the last `0.15.x` today and then we can merge the queue.

**Pull request merged**